File: demos/test-ts/src/algorithms/leetcode/s0021_merge_two_sorted_lists.test.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function mergeTwoLists(
  list1: ListNode | null,
  list2: ListNode | null
): ListNode | null {
  const dummyHead = new ListNode(0);
  let head = dummyHead;
  while (list1 || list2) {
    if (list1 === null) {
      head.next = list2;
      break;
    }
    if (list2 === null) {
      head.next = list1;
      break;
    }
    if (list1.val <= list2.val) {
      head.next = list1;
      list1 = list1.next;
    } else {
      head.next = list2;
      list2 = list2.next;
    }
    head = head.next;
  }
  return dummyHead.next;
}

File: demos/go/algorithms/leetcode/s0021_merge_two_sorted_lists_test.go
package leetcode

import (
	"test_go/algorithms/structures"
	"testing"
)

// https://leetcode.com/problems/merge-two-sorted-lists/

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	p := dummy
	for list1 != nil || list2 != nil {
		if list2 == nil {
			p.Next = list1
			break
		}
		if list1 == nil {
			p.Next = list2
			break
		}
		if list1.Val <= list2.Val {
			p.Next = list1
			list1 = list1.Next
		} else {
			p.Next = list2
			list2 = list2.Next
		}
		p = p.Next
	}
	return dummy.Next
}
func Test0021(t *testing.T) {
	tests := []struct {
		list1, list2, ans []int
	}{
		{[]int{1, 2, 4}, []int{1, 3, 4}, []int{1, 1, 2, 3, 4, 4}},
		{[]int{}, []int{0}, []int{0}},
		{[]int{}, []int{}, []int{}},
	}
	for _, test := range tests {
		r := structures.List2Ints(mergeTwoLists(structures.Ints2List(test.list1), structures.Ints2List(test.list2)))
		if !EqualInts(r, test.ans) {
			t.Errorf("expect %v, got %v", test.ans, r)
		}
	}
}
